### PR TITLE
pipeline filtering

### LINKED ISSUE #2457: Leverage path filtering for CI builds
**Description:**
To decouple frontend, backend CI builds, we can utilize path filtering to only build changed directories -- not building everything for a simple document change.

**Acceptance Criteria:** 
- [ ] Must work correct in HHS CircleCI
- [ ] Build trigger is service-specific (frontend, backend)
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
- [ ] For agnostic files (scripts/deploy.sh), ensure CI builds all
- [ ] Implement `path_ignore` for documentation updates
- [ ] Implement `path` filtering for decoupling builds on fronend vs backend
- [ ] Document new flows in README
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**
 - We are trying to improve readability and dev experience in regards to these pipelines.

**Supporting Documentation:**

**Open Questions:**

### PR DESCRIPTION
**Pull request by George-Hudson**
## Summary of Changes
Leverages GitHub Actions to trigger more nuanced workflows in CircleCI based upon branches and paths. Only the build and test workflows for the part of code that was changed should run.
Pull requests and changes to the pipeline should still run the full build and test workflows.
CircleCI still builds main, master, and release branches on HHS side. But doesn't trigger for dev team branches.
Pull request closes #2457 

## How to Test
Change code and push, or make a PR. Only workflows that pertain to the changed materials should get run.
- only backend tests should be run if code is changed in tdrs-backend/ files
- only frontend tests should be run if code is changed in tdrs-frontend/ files
- infrastructure pipelines should only get run when making changes to files that pertain to infrastructure
- all tests should run if scripts that builds and pipelines use get changed, or changes to the pipelines themselves
- all tests should run when a PR is ready for review 

Since this is a pipeline change, testing is difficult to test besides against the actual pipeline.
[GitHub actions ](https://github.com/raft-tech/TANF-app/actions) has been tested so that code pushed to unwanted branches does not run (main, master, release/**), but is picked up by [CircleCI](https://app.circleci.com/pipelines/github/raft-tech/TANF-app/13731)
And likewise, CircleCI shouldn't run against [non-tracked branches](https://app.circleci.com/pipelines/github/raft-tech/TANF-app/13738) but instead be kicked off by Actions for [frontend/backend specific code changes](https://app.circleci.com/pipelines/github/raft-tech/TANF-app/13740) as well as [build-all for PRs](https://app.circleci.com/pipelines/github/raft-tech/TANF-app/13739)

**on the HHS side, pipelines should not be affected**
- release/*, main, and master pushes should still build the CircleCI workflow (bypassing Actions)
tested on [forked repository](https://github.com/George-Hudson/TANF-app/branches)
tested on separate [CircleCI project](https://app.circleci.com/pipelines/github/George-Hudson/TANF-app)

## Deliverables
Should be able to see only workflows being run in circleci that pertain to the code change

### [Deliverable 1: Accepted Features](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] [**code change to tdrs/backend only triggers backend tests**]
   https://github.com/raft-tech/TANF-app/actions/runs/4874720916
   https://app.circleci.com/pipelines/github/raft-tech/TANF-app/13368
+ [x] [**code change to tdrs/frontend only triggers frontend tests**]
   https://github.com/raft-tech/TANF-app/actions/runs/4874770990
   https://app.circleci.com/pipelines/github/raft-tech/TANF-app/13370
+ [x] [**code change to scropts or pipeline changes triggers all tests**]
   https://github.com/raft-tech/TANF-app/actions/runs/4874701766
   https://app.circleci.com/pipelines/github/raft-tech/TANF-app/13366
+ [x] [**infrastructure deploys are only triggered when code managing the infrastructure changes**]
   https://github.com/raft-tech/TANF-app/actions/runs/4874876141
   https://app.circleci.com/pipelines/github/raft-tech/TANF-app/13372
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

No changes to tested code were made

### [Deliverable 3: Properly Styled Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [x] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [x] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [x] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [x] Do manual code review and manual testing detect any new security issues?
+ [x] If new issues detected, is investigation and/or remediation plan documented?

### CODE REVIEW

**Comment by George-Hudson** [comment]
@andrew-jameson This will need `build-and-test-all` to reset as the `Expected` check to pass, before merging.

**Review by elipe17 — APPROVED** [review]

File: docs/Technical-Documentation/Architecture-Decision-Record/020-pipeline-build-flow [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# 20. Pipeline Build Flow
```
Comment (jtimpe) [inline docs/Technical-Documentation/Architecture-Decision-Record/020-pipeline-build-flow]: can you make this file a `.md`?

**Review by jtimpe — APPROVED** [review]
Minor docs change suggestion but otherwise looks good to me!

File: docs/Technical-Documentation/Architecture-Decision-Record/020-pipeline-build-flow (hunk 1 above)
Comment (George-Hudson) [reply docs/Technical-Documentation/Architecture-Decision-Record/020-pipeline-build-flow]: Oops, yeah that was the intent. TY for the catch

Comment (George-Hudson) [reply docs/Technical-Documentation/Architecture-Decision-Record/020-pipeline-build-flow]: done

**Review by raftmsohani — APPROVED** [review]

**Review by andrew-jameson — APPROVED** [review]
Note to self/Alex: I will need to update workflow check for `build-and-test-all` prior to merging to develop but not before.

**Comment by ADPennington** [comment]
@George-Hudson @andrew-jameson is this ready for review? If so, can you tag me when the branch is up to date with `develop`?

**Comment by andrew-jameson** [comment]
> @George-Hudson @andrew-jameson is this ready for review? If so, can you tag me when the branch is up to date with `develop`?

@George-Hudson I have merged in latest from `develop` -- if Alex is to look at it, are we good to make the switchover to `build-and-test-all` now?

**Review by ADPennington — APPROVED** [review]
thanks @George-Hudson 👍🏾

**Pull request merged**